stackslib/src/net/api/get_tenures_fork_info.rs

```diff
@@ -0,0 +1,361 @@
+use stacks_common::types::chainstate::{
+    BlockHeaderHash, BurnchainHeaderHash, ConsensusHash, SortitionId, StacksBlockId,
+};
+use stacks_common::types::net::PeerHost;
+use stacks_common::types::StacksEpochId;
+use stacks_common::util::hash::{to_hex, Hash160};
+use stacks_common::util::HexError;
+use {serde, serde_json};
+
+use crate::chainstate::burn::db::sortdb::SortitionDB;
+use crate::chainstate::burn::BlockSnapshot;
+use crate::chainstate::nakamoto::{NakamotoBlock, NakamotoChainState, NakamotoStagingBlocksConn};
+use crate::chainstate::stacks::db::StacksChainState;
+use crate::chainstate::stacks::Error as ChainError;
+use crate::net::api::getblock_v3::NakamotoBlockStream;
+use crate::net::http::{
+    parse_bytes, parse_json, Error, HttpBadRequest, HttpChunkGenerator, HttpContentType,
+    HttpNotFound, HttpRequest, HttpRequestContents, HttpRequestPreamble, HttpResponse,
+    HttpResponseContents, HttpResponsePayload, HttpResponsePreamble, HttpServerError, HttpVersion,
+};
+use crate::net::httpcore::{
+    HttpRequestContentsExtensions, RPCRequestHandler, StacksHttp, StacksHttpRequest,
+    StacksHttpResponse,
+};
+use crate::net::{Error as NetError, StacksNodeState, TipRequest, MAX_HEADERS};
+use crate::util_lib::db::{DBConn, Error as DBError};
+
+pub static RPC_TENURE_FORKING_INFO_PATH: &str = "/v3/tenures_fork_info";
+
+static DEPTH_LIMIT: usize = 10;
+
+/// Struct for information about a tenure that is used to determine whether
+///  or not the tenure should have been validly forked.
+#[derive(PartialEq, Debug, Clone, Serialize, Deserialize)]
+pub struct TenureForkingInfo {
+    /// The burnchain header hash of the block that triggered this event.
+    #[serde(with = "prefix_hex")]
+    pub burn_block_hash: BurnchainHeaderHash,
+    /// The burn height of the block that triggered this event.
+    pub burn_block_height: u64,
+    /// This sortition ID of the block that triggered this event. This incorporates
+    ///  PoX forking information and the burn block hash to obtain an identifier that is
+    ///  unique across PoX forks and burnchain forks.
+    #[serde(with = "prefix_hex")]
+    pub sortition_id: SortitionId,
+    /// The parent of this burn block's Sortition ID
+    #[serde(with = "prefix_hex")]
+    pub parent_sortition_id: SortitionId,
+    /// The consensus hash of the block that triggered this event. This incorporates
+    ///  PoX forking information and burn op information to obtain an identifier that is
+    ///  unique across PoX forks and burnchain forks.
+    #[serde(with = "prefix_hex")]
+    pub consensus_hash: ConsensusHash,
+    /// Boolean indicating whether or not there was a succesful sortition (i.e. a winning
+    ///  block or miner was chosen).
+    pub was_sortition: bool,
+    /// If the sortition occurred, and a block was mined during the tenure, this is the
+    /// tenure's block.
+    #[serde(with = "prefix_opt_hex")]
+    pub first_block_mined: Option<StacksBlockId>,
+}
+
+mod prefix_opt_hex {
```

Comment:
Same comment as before. Would be nice to have unit tests for the serde functions.

Reply:
Added in fdcaea8b91972ad721d9d9e0579ac89fb531ddc0 (stackslib/src/net/api/tests/mod.rs)